pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+				if flush > half {
+					flush = half
+				}
+			}
+		}
+	}
+	return time.Duration(flush) * time.Second
+}
+
+func (a *TokenTimeoutUpdater) clientTimeout(name string) int32 {
+	t, ok := a.clientTimeouts[name]
+	if !ok {
+		c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				glog.Info("Token timeout set but OAuthClient %s not found!", name)
+			} else {
+				glog.Error("Failed to fetch OAuthClient for timeout value")
+			}
+			return 0
+		}
+		if c.AccessTokenTimeoutSeconds == nil {
+			glog.Info("Token timeout set but OAuthClient %s has no timeout!", name)
+			return 0
+		}
+		t = *c.AccessTokenTimeoutSeconds
+		a.clientTimeouts[name] = t
+	}
+	return t
+}
+
+func (a *TokenTimeoutUpdater) insert(td tokenData) {
+	a.data[td.name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.name, td.timeout})
+}
+
+func (a *TokenTimeoutUpdater) remove(td tokenData) {
+	a.tree.Delete(&tokenDataRef{td.name, td.timeout})
+	delete(a.data, td.name)
+}
+
+func (a *TokenTimeoutUpdater) flush(flushHorizon time.Time) {
+	flushedTokens := 0
+	totalTokens := len(a.data)
+
+	glog.Infof("Flushing tokens timing out before %v", flushHorizon)
+
+	for {
+		item := a.tree.Min()
+		if item == nil {
+			// out of items
+			break
+		}
+		td := a.data[item.(*tokenDataRef).name]
+		if td.timeout.Before(flushHorizon) {
+			delta := a.clientTimeout(td.client)
+			// ignore tokens for which we can't find a delta
+			if delta > 0 {
+				token, err := a.tokens.Get(td.name, metav1.GetOptions{})
+				if err == nil { // just ignore if the token is gone
+					newtimeout := int32(td.seen.Sub(token.CreationTimestamp.Time)/time.Second) + delta
+					if newtimeout > token.TimeoutsIn {
+						token.TimeoutsIn = newtimeout
+						// ignore errors, if it fails to update there isn't much we can do
+						_, _ = a.tokens.Update(token)
+					}
+				}
+			}
+			a.remove(td)
+			flushedTokens++
+		} else {
+			// out of items within the flush Horizon
+			break
+		}
+	}
+
+	glog.Infof("Flushed %d tokens out of %d in bucket", flushedTokens, totalTokens)
+}
+
+func (a *TokenTimeoutUpdater) Start() {
+	// TODO: make safety margin configurable ?
+	var safetyMargin time.Duration = time.Second * 15
+
+	glog.Infof("Started Token Timeout Flush Handling thread!")
+
+	flushTimeout := a.updateClientTimeouts()
+	nextTimer := time.NewTimer(flushTimeout)
+	nextTimeout := time.Now().Add(flushTimeout)
+
+	for {
+		select {
+		case td, ok := <-a.tokenChannel:
+			if !ok {
+				// when channel closes terminate
+				break
+			}
+			a.insert(td)
+			// if this token is going to time out before the timer, fire
+			// immediately (safety margin is added to avoid racing too close)
+			if nextTimeout.Add(safetyMargin).After(td.timeout) {
+				glog.Infof("Timeout falls below safety margin (%s < %s) forcing flush", td.timeout, nextTimeout.Add(safetyMargin))
+				if !nextTimer.Stop() {
+					<-nextTimer.C
```

Comment:
Not sure what you are trying to do here.

Reply:
when you call nextTimer.Stop() you need to check the return.
If the return is false it means the timer already fired and you have an element wait in the channel queue, in that case you have to consume the channel.